toolchain/check/subst.h

```diff
@@ -20,6 +21,24 @@ struct Substitution {
 
 using Substitutions = llvm::ArrayRef<Substitution>;
 
+// A function that performs any needed substitution into an instruction, as part
+// of substituting into a symbolic instruction sequence. The instruction ID
+// should be updated as necessary to represent the new instruction. If the
+// function returns true, the instruction is treated as fully-substituted; if it
+// returns false, the instruction will be decomposed and substitution will be
+// performed recursively into its operands.
+using SubstInstFn = llvm::function_ref<auto(SemIR::InstId& inst_id)->bool>;
+
+// A function that rebuilds an instruction after substitution. `orig_inst_id` is
+// the instruction prior to substitution, and `new_inst` is the substituted
+// instruction. Returns the new instruction ID to use to refer to `new_inst`.
+using SubstRebuildFn = llvm::function_ref<
+    auto(SemIR::InstId orig_inst_id, SemIR::Inst new_inst)->SemIR::InstId>;
+
+// Performs substitution into `inst_id` and its operands recursively.
```

Comment:
Maybe the flow should have some more comments here? From the comments:

subst_fn: "A function that performs any needed substitution into an instruction"
rebuild_fn: "A function that rebuilds an instruction after substitution."
SubstInst: "Performs substitution into `inst_id` and its operands recursively."

Is `subst_fn` called on all the operands of `inst_id` unconditionally, or only recursively? On line 237, "// No need to rebuild this instruction.", why do instructions that return false on subst_fn but have no operands not get rebuilt (or, why are they allowed to return false?)? I've also commented on `Rebuild`, because the call to `rebuild_fn` looks condition in a way I don't see explained here. 

Here, maybe something like:

> Performs substitution into `inst_id`. On each instruction recursively processed, `subst_fn` is called; if it returns false, this recurses into the instruction's operands (if any) before finishing substitution of the instruction and calling `rebuild_fn` [when <condition>].

But it may be worth looking at further than just that.

Reply:
Did a pass over the comments.